### PR TITLE
Move from nox to tox

### PR DESCRIPTION
**Pull request by gaborbernat**
To be added on top of #47.

- add a coverage report environment
- move to src layout to fix packaging (build name is reserved at root
  and will be filtered out by https://github.com/pypa/setuptools/blob/master/setuptools/command/egg_info.py#L592)
- add documentation build environment

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+passenv = *
+deps = pre-commit >= 2
+skip_install = True
+commands =
+    pre-commit run --all-files --show-diff-on-failure
+    python -c 'import pathlib; print("hint: run \{\} install to add checks as pre-commit hook".format(pathlib.Path(r"{envdir}") / "bin" / "pre-commit"))'
+
+[testenv:dev]
+description = generate a DEV environment
+extras = testing, docs
```
Comment (sinoroc) [inline tox.ini:78]: Shouldn't it be `extras = test, docs`?

File: setup.cfg [hunk 2]
```diff
@@ -57,4 +62,21 @@ line_length = 127
...
+    */src
+    *\src
+
+[coverage:run]
+branch = true
+parallel = true
+dynamic_context = test_function
+
+[coverage:html]
+show_contexts = true
```
Comment (FFY00) [inline setup.cfg:82]: This should be in a separate PR, or at least commit.

File: setup.cfg [hunk 3]
```diff
@@ -57,4 +62,21 @@ line_length = 127
...
+known_third_party = pallets_sphinx_themes,pep517,pytest,setuptools,toml
+
+[coverage:paths]
+source =
+    src
+    .tox/*/lib/python*/site-packages
+    .tox/pypy*/site-packages
+    .tox\*\Lib\site-packages\
+    */src
+    *\src
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply setup.cfg:74]: Because we run coverage on what's actually used (e.g. installed package, and we need to map it back to the source for reporting), see https://coverage.readthedocs.io/en/coverage-5.1/config.html#paths

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+parallel_show_output = True
+
+[testenv:docs]
+basepython = python3.8
+description = build documentation
+extras = doc
+passenv = *
+commands =
+    sphinx-build -d "{envtmpdir}/doctree" docs/source "{toxworkdir}/docs_out" --color -b html {posargs}
+    python -c 'import pathlib; print("documentation available under file://\{0\}".format(pathlib.Path(r"{toxworkdir}") / "docs_out" / "index.html"))'
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: Why use the platform-dependent makefile that does not work on windows when you can use tox that works on all platforms? I want to actually remove that makefile, but thought it's out of scope of this PR.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+[testenv:fix_lint]
+description = format the code base to adhere to our styles, and complain about what we cannot do automatically
+basepython = python3.8
+passenv = *
+deps = pre-commit >= 2
+skip_install = True
+commands =
+    pre-commit run --all-files --show-diff-on-failure
+    python -c 'import pathlib; print("hint: run \{\} install to add checks as pre-commit hook".format(pathlib.Path(r"{envdir}") / "bin" / "pre-commit"))'
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:74]: Which part?

File: setup.cfg (hunk 2 above)
Comment (gaborbernat) [reply setup.cfg:82]: I guess the merge coverage part (that we did not do until now) can be a follow-up PR.

File: setup.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply setup.cfg:74]: Shouldn't `src` be enough? In my report I only have `src`.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -18,20 +18,13 @@ jobs:
...
-    - name: Install dependencies
-      run: pip install toml pep517 packaging importlib_metadata
-
-    - name: Install python2 dependencies
-      if: ${{ matrix.python == 2.7 }}
-      run: pip install typing
-
-    - name: Install build backend
-      run: pip install setuptools wheel
+    - name: install build
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:26]: I generally only capitalize stuff that wants to be a paragraph and ends with a period. This does not so IMHO it's more accurate to keep it lower case. My 2c

[the remark this answers is not in the document]
Comment (FFY00) [reply .github/workflows/build.yml:26]: Well, we can discuss the style in a separate issue/PR.

**Comment by gaborbernat** [comment]
> Moving the module to a `src` folder should be in a separate PR, and btw, I do not agree.

Can you expand on this? The package the way it is does not build, and the reason is in this PR is that once I moved to tox the issue was exposed. So had to move to src to fix the issue. In theory, could be a before PR, but not a fan of trying to make it work first with nox just so that then we move it to tox.

**Comment by FFY00** [comment]
> The package the way it is does not build

What do you mean? It builds perfectly fine here, both with `python -m build` and `python setup.py build`.

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: The makefile is usable on Windows with something like Cygwin os MinGW. If you want to remove it my answer would be to just rewrite it in python, instead of manually issuing commands in tox. This currently doesn't build the API reference.

File: tox.ini (hunk 5 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:74]: Inlining python in `python -c` to print a message. Doesn't tox have a better way to deal with this? I do not think inlining escaped python inside a tox file is a great UX. This feels very hacky.

File: setup.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply setup.cfg:74]: You're seeing src exactly because this source path mapping is in place. Remove the other lines and would not be.

**Comment by gaborbernat** [comment]
> python -m build

That only works because Python adds the CWD to the sys.path and you're picking up the local files from the root, not because the build files are installed into the repository.

> python setup.py build

Not sure about that, but ``python setup.py sdist`` passes, but then does not include the ``build`` package because https://github.com/pypa/setuptools/blob/master/setuptools/command/egg_info.py#L592 removes the build folder as it's what's also used to populate the binary metafiles. Ship the sdist and users can't use this package at all.

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: > The makefile is usable on Windows with something like Cygwin os MinGW

tox is usable without needing to install cryptic things like Cygwin or MingGW. You only need python, and you already have Python because how else would you run this project.

> If you want to remove it my answer would be to just rewrite it in python, instead of manually issuing commands in tox

Why would I do that? The whole point of tox (and nox too) is to hide away Makefile like functionalities in a cross-platform way, that's driven by config in case of tox, or the nox.py in case of nox. Writing that python file would be just rewriting tox in here. See my talk https://www.youtube.com/watch?v=SFqna5ilqig for more in-depth.

> This currently doesn't build the API reference.

True, but that's a follow-up ticket, has nothing to do with this PR.

File: tox.ini (hunk 5 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:74]: It does not. tox tries to not do too much magic that can be done with already available tools. And the easiest way to print something in a cross-platform fashion (assuming you have python) is to use the print statement in Python. I can see how at first this might seem odd, but you're the first person to complain about it. There's no better way, of course I can remove the print, but feels like provides some value 🤷

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -113,30 +99,23 @@ jobs:
...
+      run: pip install . setuptools wheel
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:115]: It's semi-related. The ``build.yaml`` was broken once I moved to the src layout because this again only worked because the CWD is auto-added to the sys.path.

File: .github/workflows/build.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:26]: Thought as much, hence why only changed the lines I needed to touch to fix the breaking CI.

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: > tox is usable without needing to install cryptic things like Cygwin or MingGW. You only need python, and you already have Python because how else would you run this project.

I was not clear enough, sorry. What I was saying is that the makefile can be used on Windows, not that it would be usable in this scenario.

> Why would I do that? The whole point of tox (and nox too) is to hide away Makefile like functionalities in a cross-platform way, that's driven by config in case of tox, or the nox.py in case of nox. Writing that python file would be just rewriting tox in here. See my talk [youtube.com/watch?v=SFqna5ilqig](https://www.youtube.com/watch?v=SFqna5ilqig) for more in-depth.

Because I want to be able to build the docs without tox.

[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: > Because I want to be able to build the docs without tox.

This is not really possible in general. Someone needs to create a virtual environment, install the dependencies and then invoke the correct command to build the documentation. In theory, you can do this by invoking this commands yourself, but tox does all that for you automatically. Otherwise, who would ensure that all docs dependencies are satisfied, and they don't pollute the global namespace?

File: tox.ini (hunk 5 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:74]: I would say printing is a common need in a tool like tox, you do it here several times. Forcing people to `python -c 'print("something")'` is a very bad UX IMO. It works, so maybe that's why nobody complained.

File: .github/workflows/build.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply .github/workflows/build.yml:115]: Then use `PYTHONPATH` intead of `pip`.

File: tox.ini (hunk 5 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:74]: tox, in general, does not want to be a wrapper around already available methods and tries to avoid doing stuff it can. Printing is one of these. Noted that this is bad UX, but that does not change the fact that there's no better way available today. And to pronounce with a tox maintainer on my head now it's unlikely to change anytime soon.

File: .github/workflows/build.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply .github/workflows/build.yml:26]: You should still adhere to the style in use.

FYI, I want the first word of the name capitalized because that's the style of the default non-optional CI steps in github actions. With your change we now have one step that does not follow the style.

https://github.com/FFY00/python-build/pull/49/checks?check_run_id=802831513

File: .github/workflows/build.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:115]: That would be a bad idea. We want to validate that things work with the installed package, not that it works on the local source tree. People, in general, will use this package as installed in package format, not by copying the build files and then hacking it on the sys.path. Hacking the sys.path is exactly that allowed for you to not notice until now that the source distribution is broken. Furthermore, installing the package itself means now you not need to install the build dependencies separately, those get pulled in automatically.

File: .github/workflows/build.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:26]: I guess I'm challenging here whoever made Github names by default 🤷

File: .github/workflows/build.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply .github/workflows/build.yml:115]: I also want to validate it works in-tree, that's how I would use it when bootstrapping.

[the remark this answers is not in the document]
Comment (gaborbernat) [reply .github/workflows/build.yml:115]: Fair, but that's then separate tests. And probably better done by moving these as integration tests. I'd suggest we make such exhaustive tests there, we can run builds both as installed and inside a clear virtual environment that has no packages within, but sets the python path accordingly.

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: Me as a distro packager. I want full reproducibility, so I take care of the environment and build the package, including the docs. This is not possible with a virtual env.

Usually what you want to do is make everything usable without a virtual env, and then use it inside a virtual env if you want. It is not your call if I get to build the docs, or do anything else, inside a virtual env or not. This is the kind of mentality that makes the job significantly harder for distro packagers, or any other kind of people that need more fine-tuned control.

**Comment by FFY00** [comment]
> That only works because Python adds the CWD to the sys.path and you're picking up the local files from the root, not because the build files are installed into the site-packages of your host python.

That is not the point.

> Not sure about that, but `python setup.py sdist` passes, but then does not include the `build` package because [pypa/setuptools:setuptools/command/egg_info.py@`master`#L592](https://github.com/pypa/setuptools/blob/master/setuptools/command/egg_info.py?rgh-link-date=2020-06-24T16%3A29%3A56Z#L592) removes the build folder as it's what's also used to populate the binary metafiles. Ship the sdist and users can't use this package at all.

You're right. The wheel works though.

**Comment by gaborbernat** [comment]
> You're right. The wheel works though.

Both needs though.

> That is not the point.

What is then?

**Comment by pganssle** [comment]
FWIW I'm also +1 on a `src/` layout, which is also something we've got a rough consensus to be promoting as the default in the PyPA, and seems _especially_ germane for this project, since there's a name conflict with the location that `setuptools` currently uses to do builds in.

Good background for other reasons to structure it like this:

- Hynek Schlawack's [Testing & Packaging](https://hynek.me/articles/testing-packaging/) blog post.
- Ionel Mărieș' [Packaging a Python Library](https://blog.ionelmc.ro/2014/05/25/python-packaging/) (which is increasingly out of date for actual recommendations, but still makes a reasonable case for the `src/` layout).

If we don't use a `src/` layout, we should probably do some variation on the setup I describe in [Testing your python package as installed](https://blog.ganssle.io/articles/2019/08/test-as-installed.html).

I don't care if it's a separate PR or this PR, but given that it's _required_ if we want to continue building with `setuptools`, we should almost certainly do it.

I'm inclined to say do it as a separate PR and rebase this one on that PR, considering @gaborbernat is all about the squash merge.

**Comment by FFY00** [comment]
It's not a big deal, but should be in a separate PR.

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: > Me as a distro packager. I want full reproducibility, so I take care of the environment and build the package, including the docs. This is not possible with a virtual env.

Why wouldn't be possible? tox will ensure all that given you provide a Python to use.

> It is not your call if I get to build the docs, or do anything else, inside a virtual env or not.

I agree with the fact that you run within a virtual environment or you bring your own Python requires flexibility, and should probably be an option. However, I don't agree with the fact that the project should not control how and what packages to the provision to perform its test suite. There's a better way though here, read https://discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077 for a proposal made by Fedora on how to handle this in the long run. It also introduces https://github.com/fedora-python/tox-current-env#tox-current-env that allows running tox on a global python, rather than virtual environment where you can bring your own dependencies.

> This is the kind of mentality that makes the job significantly harder for distro packagers

Think of the Windows package managers now. Forcing a non-cross platform makefile (the fact that you can run it within cygwin does not matter, that project is a community best effort and not supported by MS itself) on them makes life significantly harder for them.

[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: > Why wouldn't be possible? tox will ensure all that given you provide a Python to use.

Reproducibility, I want to build a package today and be able to build it again in 5 years and get the *exact* same file. This is how you provide trust to the users, they do not have to rely on you saying that there is no backdoor, they can go ahead build the package themselves and get the same binary result. The language packaging (via PyPI, pip, ect.) is not there yet, but I hope it will be in the future. I would recommend you to take a look at https://reproducible-builds.org/. Installing via pip is a very big no-no for reproducibility.

If I test inside a virtual this also has absolutely no meaning what so ever for the distro, I am not testing if the package works properly with the distro packages, I am testing if the package works properly with the dependencies I fetched from PyPI and installed with pip.

> However, I don't agree with the fact that the project should not control how and what packages to the provision to perform its test suite.

Then we disagree. I have a *very* strong opinion here. I should be able to run everything in my environment, tox should just automate that. tox can make arbitrary choices on how to do it, but I should still be able to run pytest, mypy, sphinx, etc. manually.

> There's a better way though here, read [discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077](https://discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077) for a proposal made by Fedora on how to handle this in the long run. It also introduces [fedora-python/tox-current-env#tox-current-env](https://github.com/fedora-python/tox-current-env#tox-current-env) that allows running tox on a global python, rather than virtual environment where you can bring your own dependencies.

That is great. But still has some shortcommings.

> An attempt to run this with a Python version that doesn't match will fail (if tox is invoked from an Python 3.7 environment, any non 3.7 testenv will fail).

Pulling tox, the plugin and their dependencies also means that it takes more time to build packages, but I'll have to shut up and live that.

One of my main issues is that tox is not designed to run this way, it is entirely built around the idea of virtual environments. When you try to do something like this, you hit issues such as this one. I am not saying it is impossible to work around that, but it is very hard to do it properly. There is probably a reason this option is provided by a plugin and not by the upstream, right?

Right now, tox is not a tool I would want to use in distro packaging. Forcing its use a big no from me.

> Think of the Windows package managers now. Forcing a non-cross platform makefile (the fact that you can run it within cygwin does not matter, that project is a community best effort and not supported by MS itself) on them makes life significantly harder for them.

That is why I want to rewrite the makefile in Python, so that it is cross platform.

[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: > tox can make arbitrary choices on how to do it, but I should still be able to run pytest, mypy, sphinx, etc. manually.

tox wants to be the space where you can define what dependencies you request and how (more importantly what) you run. E.g. tox defines that use pytest to run the test suite and also contains the flags needed to pass in. The part that creates the environment to run that command into is configurable (and will be even further once tox 4 lands).

> That is why I want to rewrite the makefile in Python, so that it is cross platform.

tox wants to be this cross-platform tool though? Why rewrite tox when you can just use tox (which is also just pure python).

> One of my main issues is that tox is not designed to run this way, it is entirely built around the idea of virtual environments.

This is a false statement. Most people have this picture about tox mainly because that is what tox does by default, however, it's not true. We did not design tox with this enforcement in mind. By default, it uses virtual environments to achieve isolation, and pip to install dependencies. However, there's a reason both of these can be overridden entirely by plugins. This is not the only way though, through plugins you can customize both: e.g. use pacman to get python, translate PyPI dependencies to host dependencies, install them via pacman; and then ask tox to invoke the test commands, report the outcome. Each OS/distribution can create its own tox plugin and use its own tools to create a test environment, install dependencies, however the core logic of what are the dependencies, how to run them is left to drive for tox. 

> There is probably a reason this option is provided by a plugin and not by the upstream, right?

It's because the defaults are catering for 90% of people, what most people would want to use. tox does not want to dictate how to create environments/install dependencies for various distributions, so the behaviour is left up to the target dependency entirely that can be provided by plugins. This allows these plugins to evolve without needing input from tox maintainers.

[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: That is fair, but as it stands tox is undesirable for distro packagers like me. We should not enforce it's use like this.

[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: I'm not following why is less desirable than makefile.

[the remark this answers is not in the document]
Comment (FFY00) [reply tox.ini:64]: Because you can easily run the makefile without worrying about the environment. Right now for tox I need to use an external plugin which doesn't even work good enough for me.

> An attempt to run this with a Python version that doesn't match will fail (if tox is invoked from an Python 3.7 environment, any non 3.7 testenv will fail).

The main blocker is that the plugin does not simply skip if the environment is not compatible.

[the remark this answers is not in the document]
Comment (gaborbernat) [reply tox.ini:64]: How do all other projects using tox (this includes the majority of PyPa projects) manage? 

Anyways I give up. This project requires too much arguing for my side for adopting current best practices, so I resign from wanting to maintain it and focus on other projects.
